### ISSUE TITLE
Recurring actions admin view bug

### ISSUE THREAD

**Issue by crstauf** [Prospress/action-scheduler#256]
Okay, I was 50/50 on whether I was doing something wrong or if this is a bug. 

Here's what I've got:

```php
add_action( 'init', function() {
	if ( !as_next_scheduled_action( 'recurring' ) )
		as_schedule_recurring_action( strtotime( '+1 minute' ), MINUTE_IN_SECONDS, 'recurring' );
} );
```

After one run, here are the events ([screenshot](https://user-images.githubusercontent.com/4573033/54089824-9567fa00-4343-11e9-9495-066b5218c84e.png)):

|Hook|Status|Scheduled Date|Log|
|---|---|---|---|
|recurring|Complete|2019-03-10 18:46:36 +0000|**2019-03-10 18:45:36 +0000**<br />action created<br />**2019-03-10 18:46:44 +0000**<br />action started<br />**2019-03-10 18:46:44 +0000**<br />action complete|
|recurring|Pending|2019-03-10 18:46:36 +0000|**2019-03-10 18:46:44 +0000**<br />action created|

As you can see, the scheduled timestamp is the same for both events.

I am relieved to say though, that this appears to be just an admin view bug, and not a functionality bug; after a few minutes, the events do run once a minute, and continue to recur, it's just the scheduled date is the scheduled time of the initial event:

<img width="1488" alt="Screen Shot 2019-03-10 at 2 56 44 PM" src="https://user-images.githubusercontent.com/4573033/54089924-c4cb3680-4344-11e9-9016-4109c1aa341b.png">

**Comment by crstauf** [comment]
## Recurring Events
I'm registering a recurring event like-so:

```php
as_schedule_recurring_event( time(), DAY_IN_SECONDS, 'test' );
```

After one run, here's the list table:

|Hook|Status|Recurrence|Scheduled Date|Log|
|---|---|---|---|---|
|test|Complete|Every 1 day|2019-03-14 00:55:26 +0000|1. **2019-03-14 00:55:16 +0000**<br />action created<br />2. **2019-03-14 00:55:32 +0000**<br />action started<br />3. **2019-03-14 00:55:32 +0000**<br />action created|
|test|Pending|Every 1 day|2019-03-14 00:55:26 +0000|1. **2019-03-14 00:55:32 +0000**<br />action created|

What would we expect the _Scheduled Date_ column of the second row to be? I would expect the first to be as it is, with the second row being "2019-03-15 00:55:32" (that's what shows in the database), but that's clearly not what's showing.

## Cron-like Schedules
The same thing happens with cron-like schedules:

```php
as_schedule_cron_action( time() + 10, '5 1 * * * *', 'test' );
```

|Hook|Status|Recurrence|Scheduled Date|Log|
|---|---|---|---|---|
|test|Complete|Cron 5 1 * * * *|2019-03-14 01:05:00 +0000|1. **2019-03-14 01:03:09 +0000**<br />action created<br />2. **2019-03-14 01:05:05 +0000**<br />action started<br />3. **2019-03-14 01:05:05 +0000**<br />action complete|
|test|Pending|Cron 5 1 * * * *|2019-03-14 01:05:00 +0000|1. **2019-03-14 01:05:05 +0000**<br />action created|

I would expect the "Scheduled Date" column of the first row to read as it is, and the second row to read "2019-03-15 01:05:05 +0000" (matches the database).

From what I can tell, this issue stems from `ActionScheduler_IntervalSchedule::next()` and `ActionScheduler_CronSchedule::next()`. The completed action's timestamp seems to be registering just slightly before the scheduled time (like two seconds before), and so the `next()` function actually returns the first event. I'm not sure right now how to resolve that issue, but am happy to try, and can certainly add some logic to check the status of the event so the scheduled date of completed actions shows rather than the scheduled date of the next action.

@thenbrent @JPry Please review and provide feedback. Thanks.

**Comment by thenbrent** [comment]
>  I would expect the first to be as it is, with the second row being "2019-03-15 00:55:32" 

I would expect the same thing - the date to show the _next_ schedule date, not the originally scheduled date.

> The completed action's timestamp seems to be registering just slightly before the scheduled time (like two seconds before)

That _should_ be impossible...

The logs also show the action as being completed at 5 seconds after the scheduled time: `2019-03-14 01:05:05 +0000`.

I could potentially see this being an issue if timezones are at play, but the dates should all be UTC.

> I'm not sure right now how to resolve that issue, but am happy to try

If you can take a look that would be greatly appreciated. The first place I'd look into is `ActionScheduler_Abstract_QueueRunner::schedule_next_instance()`, as that is responsible for scheduling the next instance of a recurring action.

**Comment by crstauf** [comment]
@thenbrent It does appear like the events are being scheduled correctly, they're just not showing correctly in the list table.

<img width="1505" alt="Screen Shot 2019-03-14 at 12 27 37 AM" src="https://user-images.githubusercontent.com/4573033/54331349-04697b00-45f0-11e9-8e42-197a7d1aa0e2.png">

**Comment by crstauf** [comment]
This is interesting:

```
[14-Mar-2019 04:13:30 UTC] ActionScheduler_IntervalSchedule Object
(
    [start:ActionScheduler_IntervalSchedule:private] => ActionScheduler_DateTime Object
        (
            [utcOffset:protected] => 0
            [date] => 2019-03-14 04:07:20.000000
            [timezone_type] => 1
            [timezone] => +00:00
        )

    [start_timestamp:ActionScheduler_IntervalSchedule:private] => 1552536440
    [interval_in_seconds:ActionScheduler_IntervalSchedule:private] => 180
)

[14-Mar-2019 04:13:30 UTC] ActionScheduler_IntervalSchedule Object
(
    [start:ActionScheduler_IntervalSchedule:private] => ActionScheduler_DateTime Object
        (
            [utcOffset:protected] => 0
            [date] => 2019-03-14 04:07:20.000000
            [timezone_type] => 1
            [timezone] => +00:00
        )

    [start_timestamp:ActionScheduler_IntervalSchedule:private] => 1552536440
    [interval_in_seconds:ActionScheduler_IntervalSchedule:private] => 180
)

[14-Mar-2019 04:13:30 UTC] ActionScheduler_IntervalSchedule Object
(
    [start:ActionScheduler_IntervalSchedule:private] => ActionScheduler_DateTime Object
        (
            [utcOffset:protected] => 0
            [date] => 2019-03-14 04:07:20.000000
            [timezone_type] => 1
            [timezone] => +00:00
        )

    [start_timestamp:ActionScheduler_IntervalSchedule:private] => 1552536440
    [interval_in_seconds:ActionScheduler_IntervalSchedule:private] => 180
)
```

This is from `debug.log` when adding `error_log( print_r( $schedule, true ) );` to `ActionScheduler_ListTable::get_schedule_display_string()`. Is the start time supposed to be the same for all of them?

**Comment by crstauf** [comment]
@thenbrent Submitted PR #262. I am 100% sure that the changes will need to be rewritten, but the correct date and time does now show on the list table for recurring events.

**Comment by thenbrent** [comment]
I'm getting into this now @crstauf. The display issue is a surface part of a mucher deeper issue (which you already discovered based on your patches in PR #262).

I've opened #330 to address that issue. The final patch for that will fix this display bug, along with a number of other related issues.

**Comment by thenbrent** [comment]
Fixed with #333.
